### PR TITLE
CommonJS support in tgpu-cli and other improvements

### LINKED ISSUE #407: Add an option to generate CommonJS with tgpu-cli

### PR DESCRIPTION
**Pull request by mhawryluk**
closes #407

### CODE REVIEW

File: packages/tgpu-cli/gen.mjs [hunk 1]
```diff
@@ -7,56 +7,73 @@ const cwd = new URL(`file:${process.cwd()}/`);
...
 const LENGTH_VAR = 'arrayLength';
 
+let exportsList = [];
+
 /**
  * @param { string } input
  * @param { string } output
+ * @param { boolean } toTs
+ * @param { boolean } toCjs
  */
```
Comment (reczkok) [inline packages/tgpu-cli/gen.mjs]: Maybe it would be better to replace this with an enum?
```suggestion
/**
 * @enum {string}
 */
const Target = {
  TS: 'ts',
  CJS: 'cjs',
};

/**
 * @param { string } input
 * @param { string } output
 * @param { Target } target
 */
```

Comment (mhawryluk) [reply packages/tgpu-cli/gen.mjs]: don't think so, these flags are not mutually exclusive

File: packages/tgpu-cli/tgpu-cli.mjs [hunk 2]
```diff
@@ -11,18 +14,24 @@ const args = arg({
...
   '--input': String,
   '--output': String,
+  '--commonjs': Boolean,
+  '--js': Boolean,
 
   '-v': '--version',
   '-h': '--help',
   '-i': '--input',
   '-o': '--output',
+  '-c': '--commonjs',
```
Comment (iwoplaza) [inline packages/tgpu-cli/tgpu-cli.mjs]: Imo, we should not give a shortcut for `--commonjs`, hard to interpret `c` as `commonjs` when somebody first sees it.

```suggestion
```

File: packages/tgpu-cli/tgpu-cli.mjs [hunk 3]
```diff
@@ -11,18 +14,24 @@ const args = arg({
   '--help': Boolean,
   '--input': String,
   '--output': String,
+  '--commonjs': Boolean,
+  '--js': Boolean,
```
Comment (iwoplaza) [inline packages/tgpu-cli/tgpu-cli.mjs]: Could we provide a way to configure the output extension, and infer whether to generate TypeScript or JavaScript from that? Just added a task for it #427 

```suggestion
  '--ext': String,
```

File: packages/tgpu-cli/tgpu-cli.mjs [hunk 4]
```diff
@@ -21,32 +27,52 @@ const args = arg({
...
       const watch = args['--watch'] ?? false;
 
-      if (!input) {
+      if (!input || !output) {
         console.error(
-          `${color.Red}Error: Missing required argument: ${color.Yellow}--input${color.Reset}`,
+          `${color.Red}Error: Missing some of the required arguments: ${color.Yellow}--input, --output${color.Reset}`,
         );
         exit(1);
       }
```
Comment (reczkok) [inline packages/tgpu-cli/tgpu-cli.mjs]: We could be more specific with the error.

Comment (reczkok) [inline packages/tgpu-cli/tgpu-cli.mjs:61]: We could be more specific with the error.

File: packages/tgpu-cli/gen.mjs [hunk 5]
```diff
@@ -1,64 +1,107 @@
...
+    .access(output)
+    .then(() => true)
+    .catch(() => false);
+
+  if (fileExists && !overwriteExisting) {
+    console.warn(
+      `File ${output} already exists. Use --overwrite option to replace existing files.`,
+    );
+    return;
+  }
```
Comment (reczkok) [inline packages/tgpu-cli/gen.mjs]: This interacts with watch mode in a non-intuitive way. I think maybe watch mode should implicitly have the `--overwrite` flag?

File: packages/tgpu-cli/tgpu-cli.mjs [hunk 6]
```diff
@@ -21,32 +27,52 @@ const args = arg({
...
-  --watch, -w   Watch for changes in the input file(s) and regenerate the output file(s)`),
+  --input, -i\t The input file or glob pattern.
+  --output, -o\t The output name or pattern for generated file(s).
+  --watch, -w\t Watch for changes in the input file(s) and regenerate the output file(s).
+  --commonjs\t Generate a CommonJS style file.
+`),
+
     execute: async () => {
       const input = args['--input'] ?? args._[1];
       const output = args['--output'];
```
Comment (reczkok) [inline packages/tgpu-cli/tgpu-cli.mjs]: Now that output is not optional maybe we could also make it a positional argument for _convenience_.

File: packages/tgpu-cli/gen.mjs (hunk 5 above)
Comment (mhawryluk) [reply packages/tgpu-cli/gen.mjs]: I'm afraid it's quite dangerous to just treat watch mode as having the overwrite flag. While overwriting is expected when editing source file, it isn't after we just booted the script. Imo there are 2 options: add an immediate warning when someone runs the script with `watch`, but with no `overwrite`; ooor in that case only check if we're overwriting the first time we generate a file, not when the file changes. what do you think?

File: packages/tgpu-cli/tgpu-cli.mjs (hunk 6 above)
Comment (mhawryluk) [reply packages/tgpu-cli/tgpu-cli.mjs]: wouldn't that be confusing which argument should come first?

File: packages/tgpu-cli/gen.mjs (hunk 5 above)
Comment (reczkok) [reply packages/tgpu-cli/gen.mjs]: I think i like the second approach better. It is how I would assume `watch` works by default.

Comment (iwoplaza) [reply packages/tgpu-cli/gen.mjs]: Same here, second approach makes the most sense to me.

File: packages/tgpu-cli/tgpu-cli.mjs (hunk 6 above)
Comment (reczkok) [reply packages/tgpu-cli/tgpu-cli.mjs]: I don't think so but i'm not 100% on that. Maybe if someone is not sure he can use `--help`. I will take a look how other cli's handle that and get back to you.

File: packages/tgpu-cli/gen.mjs (hunk 5 above)
Comment (iwoplaza) [reply packages/tgpu-cli/gen.mjs]: I would still lean towards having both modes be explicit (either keep existing files, or overwrite existing files), and by default, exit with an error if a matching file already exists. In that way, running `watch` a second time will result in a proper error, not just a set of warnings for existing files. Then the user can choose whether they want to `--overwrite` or `--keep`.

File: packages/tgpu-cli/tgpu-cli.mjs (hunk 6 above)
Comment (reczkok) [reply packages/tgpu-cli/tgpu-cli.mjs]: It looks like what I proposed isn't common practice. I still think we should provide a way to shorten the command (like defaulting to `.wgsl -> .ts` like before). The command to produce a single file (which would be a common use case imo) is a bit wordy.

Comment (mhawryluk) [reply packages/tgpu-cli/tgpu-cli.mjs]: now that I think of it I also believe people may expect that typing multiple file names would result in all of them being inputs, not that the second one is the output

File: packages/tgpu-cli/gen.mjs [hunk 7]
```diff
@@ -1,64 +1,107 @@
...
+
 /**
  * @param { string } input
  * @param { string } output
+ * @param { boolean } toTs
+ * @param { boolean } toCommonJs
+ * @param { boolean } overwriteExisting
  */
-async function main(input, output) {
+async function main(input, output, toTs, toCommonJs, overwriteExisting) {
```
Comment (iwoplaza) [inline packages/tgpu-cli/gen.mjs]: Tiny nitpick, not fond of this many primitive parameters being passed in as positional parameters. Maybe we could merge the Context and Options into one, and pass that around instead? (this would also get rid of the class in favor of a plain object).

```ts
/**
 * @typedef {object} Options
 * @prop {string} input
 * @prop {string} output
 * @prop {boolean} toTs
 * @prop {'commonjs'|'esmodule'} moduleSyntax
 * @prop {'keep'|'overwrite'} [existingFileStrategy]
 */
```

```suggestion
/**
 * @param {Options} options
 */
async function main(options) {
```

File: packages/tgpu-cli/gen.mjs [hunk 8]
```diff
@@ -1,64 +1,107 @@
...
   const inputPath = new URL(input, cwd);
   const outputPath = new URL(output, cwd);
-  const toTs = output.endsWith('.ts');
   const inputContents = await fs.readFile(inputPath, 'utf8');
 
-  await fs.writeFile(outputPath, generate(inputContents, toTs));
+  const generated = generate(inputContents, new Context({ toTs, toCommonJs }));
```
Comment (iwoplaza) [inline packages/tgpu-cli/gen.mjs]:
```suggestion
  const generated = generate(inputContents, options);
```

File: packages/tgpu-cli/tgpu-cli.mjs [hunk 9]
```diff
@@ -21,32 +27,53 @@ const args = arg({
...
 
 Options:
-  --input, -i   The input file or glob pattern.
-  --output, -o  The output file. If not provided, the input file will be used with a .ts extension. Cannot be used with glob patterns.
-  --watch, -w   Watch for changes in the input file(s) and regenerate the output file(s)`),
+  --input, -i\t The input file or glob pattern.
+  --output, -o\t The output name or pattern for generated file(s).
+  --watch, -w\t Watch for changes in the input file(s) and regenerate the output file(s).
+  --commonjs\t Generate a CommonJS style file.
+  --overwrite\t Force overwriting existing files.
```
Comment (iwoplaza) [inline packages/tgpu-cli/tgpu-cli.mjs]:
```suggestion
  --overwrite\t Overwrite existing files.
  --keep\t Keep existing files.
```

**Review by iwoplaza — CHANGES_REQUESTED** [review]

File: packages/tgpu-cli/gen.mjs [hunk 10]
```diff
@@ -1,100 +1,153 @@
...
 
 /**
- * @param { string } input
- * @param { string } output
+ * @typedef {object} Options
+ * @prop {string} inputPath
+ * @prop {string} outputPath
+ * @prop {boolean} toTs
+ * @prop {'commonjs' | 'esmodule'} moduleSyntax
+ * @prop {'keep'|'overwrite'} [existingFileStrategy]
```
Comment (reczkok) [inline packages/tgpu-cli/gen.mjs]:
```suggestion
 * @prop {'keep' | 'overwrite'} [existingFileStrategy]
```

**Review by reczkok — APPROVED** [review]
Great work! 🛠️ Really good idea with the output path tests, it was a pain to do manually💜

File: packages/tgpu-cli/tgpu-cli.mjs [hunk 11]
```diff
@@ -56,26 +103,44 @@ Options:
...
 
-      if (output && files.length > 1) {
+      const fileNames = files.map((file) => path.parse(file).name);
+      const duplicates = fileNames.filter(
+        (name, index) => fileNames.indexOf(name) !== index,
+      );
+      if (duplicates.length > 0 && !/\*\*\/.*\*.*/.test(output)) {
         console.error(
-          `${color.Red}Error: More than one file found, while a single output name was provided ${color.Reset}`,
+          `${color.Red}Error: Duplicates found with name(s): [${duplicates.join(', ')}], while a single directory output pattern was provided. Make sure your pattern contains "**/*" to keep the original directory structure. ${color.Reset}`,
```
Comment (iwoplaza) [inline packages/tgpu-cli/tgpu-cli.mjs:119]: 🤩 💛

File: packages/tgpu-cli/tgpu-cli.mjs [hunk 12]
```diff
@@ -56,26 +103,44 @@ Options:
...
+      /**
+       * @param {{ exitOnError: boolean, files: string[], checkExisting: boolean }} options
+       */
+      const processFiles = async ({ exitOnError, files, checkExisting }) => {
         const results = await Promise.allSettled(
           files.map(async (file) => {
-            const out = output ?? file.replace('.wgsl', '.ts');
-            console.log(`Generating ${file} >>> ${out}`);
-            return generate(file, out).catch((error) => {
+            const outputhPath = outputPathCompiler(file);
```
Comment (iwoplaza) [inline packages/tgpu-cli/tgpu-cli.mjs]: Tiny nit
```suggestion
            const outputPath = outputPathCompiler(file);
```

File: packages/tgpu-cli/tgpu-cli.mjs [hunk 13]
```diff
@@ -56,26 +103,44 @@ Options:
...
+
+            console.log(`Generating ${file} >>> ${outputhPath}`);
+            return generate({
+              inputPath: file,
+              outputPath: outputhPath,
```
Comment (iwoplaza) [inline packages/tgpu-cli/tgpu-cli.mjs]:
```suggestion
              outputPath,
```

**Review by iwoplaza — APPROVED** [review]
Amazing work!! 🙌 🐇
